File: router-deeplink/code/android/app/src/main/java/com/example/router/domain/validation/RouteValidator.kt
package com.example.router.domain.validation

import com.example.router.domain.model.RouteRequest

class RouteValidator {
  private val allowedRoutes = setOf("home", "explore", "profile")
  private val allowedFilters = setOf("active", "recent", "all")
  private val allowedSorts = setOf("newest", "oldest")

  fun validate(request: RouteRequest): ValidationResult {
    if (!allowedRoutes.contains(request.route)) {
      return ValidationResult(false, "unknown route")
    }

    if (request.route == "explore") {
      val filter = request.params["filter"]
      if (filter == null || !allowedFilters.contains(filter)) {
        return ValidationResult(false, "invalid filter")
      }
      val sort = request.params["sort"]
      if (sort != null && !allowedSorts.contains(sort)) {
        return ValidationResult(false, "invalid sort")
      }
    }

    if (request.route == "profile") {
      val key = request.params["toggle"]
      if (key.isNullOrBlank()) {
        return ValidationResult(false, "missing toggle key")
      }
      val value = request.params["value"]
      if (value.isNullOrBlank() || (value != "true" && value != "false")) {
        return ValidationResult(false, "invalid toggle value")
      }
    }

    return ValidationResult(true)
  }
}

data class ValidationResult(val valid: Boolean, val reason: String? = null)

File: router-deeplink/code/ios/RouterAppApp/RouterCore/Domain/Validation/RouteValidator.swift
import Foundation

final class RouteValidator {
  private let allowedRoutes: Set<String> = ["home", "explore", "profile"]
  private let allowedFilters: Set<String> = ["active", "recent", "all"]
  private let allowedSorts: Set<String> = ["newest", "oldest"]

  func validate(_ request: RouteRequest) -> ValidationResult {
    if !allowedRoutes.contains(request.route) {
      return ValidationResult(valid: false, reason: "unknown route")
    }

    if request.route == "explore" {
      guard let filter = request.params["filter"], allowedFilters.contains(filter) else {
        return ValidationResult(valid: false, reason: "invalid filter")
      }
      if let sort = request.params["sort"], !allowedSorts.contains(sort) {
        return ValidationResult(valid: false, reason: "invalid sort")
      }
    }

    if request.route == "profile" {
      let key = request.params["toggle"]
      if key == nil || key == "" {
        return ValidationResult(valid: false, reason: "missing toggle key")
      }
      let value = request.params["value"]
      if value != "true" && value != "false" {
        return ValidationResult(valid: false, reason: "invalid toggle value")
      }
    }

    return ValidationResult(valid: true, reason: nil)
  }
}

struct ValidationResult {
  let valid: Bool
  let reason: String?
}
